File: src/cpp/src/tools.cpp
/**
 * @file tools.cpp
 * @brief Tool utilities for GeminiSDK C++
 */

#include "geminisdk/tools.hpp"

namespace geminisdk {

ToolParametersBuilder& ToolParametersBuilder::add_string(const std::string& name, const std::string& description) {
    properties_[name] = {
        {"type", "string"},
        {"description", description}
    };
    return *this;
}

ToolParametersBuilder& ToolParametersBuilder::add_number(const std::string& name, const std::string& description) {
    properties_[name] = {
        {"type", "number"},
        {"description", description}
    };
    return *this;
}

ToolParametersBuilder& ToolParametersBuilder::add_integer(const std::string& name, const std::string& description) {
    properties_[name] = {
        {"type", "integer"},
        {"description", description}
    };
    return *this;
}

ToolParametersBuilder& ToolParametersBuilder::add_boolean(const std::string& name, const std::string& description) {
    properties_[name] = {
        {"type", "boolean"},
        {"description", description}
    };
    return *this;
}

ToolParametersBuilder& ToolParametersBuilder::add_enum(
    const std::string& name,
    const std::string& description,
    const std::vector<std::string>& values
) {
    properties_[name] = {
        {"type", "string"},
        {"description", description},
        {"enum", values}
    };
    return *this;
}

ToolParametersBuilder& ToolParametersBuilder::required(const std::vector<std::string>& fields) {
    required_ = fields;
    return *this;
}

json ToolParametersBuilder::build() const {
    return {
        {"properties", properties_},
        {"required", required_}
    };
}

Tool create_tool(
    const std::string& name,
    const std::string& description,
    const std::optional<json>& parameters
) {
    Tool tool;
    tool.name = name;
    tool.description = description;
    tool.parameters = parameters;
    return tool;
}

Tool define_tool(
    const std::string& name,
    const std::string& description,
    const ToolParametersBuilder& builder
) {
    return create_tool(name, description, builder.build());
}

Tool declarative_tool(
    const std::string& name,
    const std::optional<json>& params
) {
    Tool tool;
    tool.name = name;
    tool.description = "Tool: " + name;
    if (params.has_value() && !params->empty()) {
        tool.parameters = params;
    }
    return tool;
}

std::vector<Tool> normalize_tools_from_json(const json& specs) {
    std::vector<Tool> result;
    if (!specs.is_array()) {
        return result;
    }
    for (const auto& item : specs) {
        if (item.is_object()) {
            for (auto it = item.begin(); it != item.end(); ++it) {
                std::optional<json> params;
                if (it.value().is_object() && !it.value().empty()) {
                    params = it.value();
                }
                result.push_back(declarative_tool(it.key(), params));
            }
        }
    }
    return result;
}

ToolResult success_result(const std::string& text) {
    ToolResult result;
    result.result_type = ToolResultType::Success;
    result.text_result_for_llm = text;
    return result;
}

ToolResult failure_result(const std::string& text) {
    ToolResult result;
    result.result_type = ToolResultType::Failure;
    result.text_result_for_llm = text;
    return result;
}

ToolResult rejected_result(const std::string& text) {
    ToolResult result;
    result.result_type = ToolResultType::Rejected;
    result.text_result_for_llm = text;
    return result;
}

void ToolRegistry::register_tool(const Tool& tool, ToolHandler handler) {
    std::lock_guard<std::mutex> lock(mutex_);
    tools_[tool.name] = tool;
    handlers_[tool.name] = handler;
}

std::vector<Tool> ToolRegistry::tools() const {
    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<Tool> result;
    for (const auto& [name, tool] : tools_) {
        result.push_back(tool);
    }
    return result;
}

std::optional<Tool> ToolRegistry::get_tool(const std::string& name) const {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = tools_.find(name);
    if (it != tools_.end()) {
        return it->second;
    }
    return std::nullopt;
}

ToolHandler ToolRegistry::get_handler(const std::string& name) const {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = handlers_.find(name);
    if (it != handlers_.end()) {
        return it->second;
    }
    return nullptr;
}

ToolResult ToolRegistry::execute(const ToolInvocation& invocation) {
    ToolHandler handler;
    {
        std::lock_guard<std::mutex> lock(mutex_);
        auto it = handlers_.find(invocation.name);
        if (it == handlers_.end()) {
            return failure_result("Tool '" + invocation.name + "' not found");
        }
        handler = it->second;
    }
    return handler(invocation);
}

void ToolRegistry::unregister(const std::string& name) {
    std::lock_guard<std::mutex> lock(mutex_);
    tools_.erase(name);
    handlers_.erase(name);
}

bool ToolRegistry::has(const std::string& name) const {
    std::lock_guard<std::mutex> lock(mutex_);
    return tools_.find(name) != tools_.end();
}

std::vector<std::string> ToolRegistry::names() const {
    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<std::string> result;
    for (const auto& [name, _] : tools_) {
        result.push_back(name);
    }
    return result;
}

} // namespace geminisdk

File: src/rust/src/tools.rs
//! Tool utilities for GeminiSDK Rust.

use crate::types::{Tool, ToolInvocation, ToolResult, ToolResultType};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

pub type BoxedToolHandler =
    Arc<dyn Fn(ToolInvocation) -> Pin<Box<dyn Future<Output = ToolResult> + Send>> + Send + Sync>;

/// Creates a tool definition from a name, description, and parameters schema.
pub fn create_tool(
    name: impl Into<String>,
    description: impl Into<String>,
    parameters: Option<Value>,
) -> Tool {
    Tool {
        name: name.into(),
        description: description.into(),
        parameters,
    }
}

/// A macro-friendly helper to define tool parameters.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ToolParameters {
    #[serde(default)]
    pub properties: HashMap<String, ToolProperty>,
    #[serde(default)]
    pub required: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolProperty {
    #[serde(rename = "type")]
    pub prop_type: String,
    pub description: Option<String>,
    #[serde(rename = "enum", skip_serializing_if = "Option::is_none")]
    pub enum_values: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default: Option<Value>,
}

impl ToolParameters {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_string(mut self, name: impl Into<String>, description: impl Into<String>) -> Self {
        self.properties.insert(
            name.into(),
            ToolProperty {
                prop_type: "string".to_string(),
                description: Some(description.into()),
                enum_values: None,
                default: None,
            },
        );
        self
    }

    pub fn add_number(mut self, name: impl Into<String>, description: impl Into<String>) -> Self {
        self.properties.insert(
            name.into(),
            ToolProperty {
                prop_type: "number".to_string(),
                description: Some(description.into()),
                enum_values: None,
                default: None,
            },
        );
        self
    }

    pub fn add_integer(mut self, name: impl Into<String>, description: impl Into<String>) -> Self {
        self.properties.insert(
            name.into(),
            ToolProperty {
                prop_type: "integer".to_string(),
                description: Some(description.into()),
                enum_values: None,
                default: None,
            },
        );
        self
    }

    pub fn add_boolean(mut self, name: impl Into<String>, description: impl Into<String>) -> Self {
        self.properties.insert(
            name.into(),
            ToolProperty {
                prop_type: "boolean".to_string(),
                description: Some(description.into()),
                enum_values: None,
                default: None,
            },
        );
        self
    }

    pub fn add_enum(
        mut self,
        name: impl Into<String>,
        description: impl Into<String>,
        values: Vec<String>,
    ) -> Self {
        self.properties.insert(
            name.into(),
            ToolProperty {
                prop_type: "string".to_string(),
                description: Some(description.into()),
                enum_values: Some(values),
                default: None,
            },
        );
        self
    }

    pub fn required(mut self, fields: Vec<&str>) -> Self {
        self.required = fields.into_iter().map(String::from).collect();
        self
    }

    pub fn to_value(&self) -> Value {
        serde_json::to_value(self).unwrap_or(json!({}))
    }
}

/// Tool registry for managing multiple tools.
pub struct ToolRegistry {
    tools: HashMap<String, Tool>,
    handlers: HashMap<String, BoxedToolHandler>,
}

impl Default for ToolRegistry {
    fn default() -> Self {
        Self::new()
    }
}

impl ToolRegistry {
    pub fn new() -> Self {
        Self {
            tools: HashMap::new(),
            handlers: HashMap::new(),
        }
    }

    /// Register a tool with its handler.
    pub fn register<F, Fut>(&mut self, tool: Tool, handler: F)
    where
        F: Fn(ToolInvocation) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = ToolResult> + Send + 'static,
    {
        let name = tool.name.clone();
        self.tools.insert(name.clone(), tool);
        self.handlers.insert(
            name,
            Arc::new(move |inv| Box::pin(handler(inv)) as Pin<Box<dyn Future<Output = ToolResult> + Send>>),
        );
    }

    /// Get all registered tools.
    pub fn tools(&self) -> Vec<Tool> {
        self.tools.values().cloned().collect()
    }

    /// Get a tool by name.
    pub fn get_tool(&self, name: &str) -> Option<&Tool> {
        self.tools.get(name)
    }

    /// Get a handler by name.
    pub fn get_handler(&self, name: &str) -> Option<&BoxedToolHandler> {
        self.handlers.get(name)
    }

    /// Execute a tool invocation.
    pub async fn execute(&self, invocation: ToolInvocation) -> ToolResult {
        if let Some(handler) = self.handlers.get(&invocation.name) {
            handler(invocation).await
        } else {
            ToolResult {
                result_type: Some(ToolResultType::Failure),
                text_result_for_llm: Some(format!("Tool '{}' not found", invocation.name)),
                binary_result: None,
                session_log: None,
            }
        }
    }

    /// Unregister a tool.
    pub fn unregister(&mut self, name: &str) {
        self.tools.remove(name);
        self.handlers.remove(name);
    }

    /// Check if a tool is registered.
    pub fn has(&self, name: &str) -> bool {
        self.tools.contains_key(name)
    }

    /// Get all tool names.
    pub fn names(&self) -> Vec<String> {
        self.tools.keys().cloned().collect()
    }
}

/// Helper to create a success result.
pub fn success_result(text: impl Into<String>) -> ToolResult {
    ToolResult {
        result_type: Some(ToolResultType::Success),
        text_result_for_llm: Some(text.into()),
        binary_result: None,
        session_log: None,
    }
}

/// Helper to create a failure result.
pub fn failure_result(text: impl Into<String>) -> ToolResult {
    ToolResult {
        result_type: Some(ToolResultType::Failure),
        text_result_for_llm: Some(text.into()),
        binary_result: None,
        session_log: None,
    }
}

/// Helper to create a rejected result.
pub fn rejected_result(text: impl Into<String>) -> ToolResult {
    ToolResult {
        result_type: Some(ToolResultType::Rejected),
        text_result_for_llm: Some(text.into()),
        binary_result: None,
        session_log: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_tool() {
        let tool = create_tool(
            "test_tool",
            "A test tool",
            Some(
                ToolParameters::new()
                    .add_string("input", "The input string")
                    .required(vec!["input"])
                    .to_value(),
            ),
        );

        assert_eq!(tool.name, "test_tool");
        assert_eq!(tool.description, "A test tool");
        assert!(tool.parameters.is_some());
    }

    #[test]
    fn test_tool_parameters() {
        let params = ToolParameters::new()
            .add_string("name", "User name")
            .add_integer("age", "User age")
            .add_boolean("active", "Is active")
            .add_enum("status", "User status", vec!["online".into(), "offline".into()])
            .required(vec!["name", "age"]);

        let value = params.to_value();
        assert!(value.get("properties").is_some());
        assert!(value.get("required").is_some());
    }
}
